packages/lodestar/src/chain/validation/attestation.ts

```diff
@@ -1,177 +1,286 @@
+
+  // [REJECT] The attestation is for the correct subnet
+  // -- i.e. compute_subnet_for_attestation(committees_per_slot, attestation.data.slot, attestation.data.index) == subnet_id,
+  // where committees_per_slot = get_committee_count_per_slot(state, attestation.data.target.epoch),
+  // which may be pre-computed along with the committee information for the signature check.
+  const expectedSubnet = computeSubnetForSlot(attestationTargetState, attSlot, attIndex);
+  if (subnet !== null && subnet !== expectedSubnet) {
     throw new AttestationError({
       code: AttestationErrorCode.INVALID_SUBNET_ID,
       received: subnet,
       expected: expectedSubnet,
-      job: attestationJob,
     });
   }
 
-  const indexedAttestation = attestationTargetState.getIndexedAttestation(attestation);
-
-  // Do verify signature
-  if (!attestationJob.validSignature) {
-    const signatureSet = allForks.getIndexedAttestationSignatureSet(attestationTargetState, indexedAttestation);
-    if (!(await chain.bls.verifySignatureSets([signatureSet]))) {
-      throw new AttestationError({
-        code: AttestationErrorCode.INVALID_SIGNATURE,
-        job: attestationJob,
-      });
-    }
+  // [IGNORE] There has been no other valid attestation seen on an attestation subnet that has an
+  // identical attestation.data.target.epoch and participating validator index.
+  if (chain.seenAttesters.isKnown(targetEpoch, validatorIndex)) {
+    throw new AttestationError({code: AttestationErrorCode.ATTESTATION_ALREADY_KNOWN, targetEpoch, validatorIndex});
   }
 
-  // verifySignature = false, verified above
-  if (!phase0.isValidIndexedAttestation(attestationTargetState, indexedAttestation, false)) {
-    throw new AttestationError({
-      code: AttestationErrorCode.INVALID_INDEXED_ATTESTATION,
-      job: attestationJob,
-    });
+  // [REJECT] The signature of attestation is valid.
+  const indexedAttestation: phase0.IndexedAttestation = {
+    attestingIndices: [validatorIndex] as List<number>,
+    data: attData,
+    signature: attestation.signature,
+  };
+  const signatureSet = getIndexedAttestationSignatureSet(attestationTargetState, indexedAttestation);
+  if (!(await chain.bls.verifySignatureSets([signatureSet]))) {
+    throw new AttestationError({code: AttestationErrorCode.INVALID_SIGNATURE});
   }
 
-  if (!ssz.Epoch.equals(attestation.data.target.epoch, computeEpochAtSlot(attestationSlot))) {
-    throw new AttestationError({
-      code: AttestationErrorCode.BAD_TARGET_EPOCH,
-      job: attestationJob,
-    });
+  // Now that the attestation has been fully verified, store that we have received a valid attestation from this validator.
+  //
+  // It's important to double check that the attestation still hasn't been observed, since
+  // there can be a race-condition if we receive two attestations at the same time and
+  // process them in different threads.
+  if (chain.seenAttesters.isKnown(targetEpoch, validatorIndex)) {
```

Comment:
this is a duplicate, checked above already

Reply:
> this is a duplicate, checked above already

The duplication is intentional, since there as async call before to the bls verification there's a race condition.

```
// It's important to double check that the attestation still hasn't been observed, since
  // there can be a race-condition if we receive two attestations at the same time and
  // process them in different threads.
```